libsawtooth/src/store/lmdb.rs

```diff
@@ -326,6 +308,123 @@ impl<
     }
 }
 
+/// `LmdbOrderedStoreIter` is an iterator over entries in an LMDB ordered store. It uses an
+/// in-memory cache of items that it retrieves from the database using a cursor. The cache is
+/// refilled when it is emptied, and its size is defined by the ITER_CACHE_SIZE constant.
+///
+/// An optional `OrderedStoreRange` may be provided to get only a subset of entries from the
+/// database.
+struct LmdbOrderedStoreIter<V, I> {
+    env: Arc<lmdb::Environment>,
+    index_to_key_db: Arc<lmdb::Database<'static>>,
+    main_db: Arc<lmdb::Database<'static>>,
+
+    cache: VecDeque<V>,
+    range: OrderedStoreRange<I>,
+}
+
+impl<V: FromBytes, I: AsBytes + FromBytes + PartialEq + PartialOrd> LmdbOrderedStoreIter<V, I> {
+    fn new(
+        env: Arc<lmdb::Environment>,
+        index_to_key_db: Arc<lmdb::Database<'static>>,
+        main_db: Arc<lmdb::Database<'static>>,
+        range: Option<OrderedStoreRange<I>>,
+    ) -> Result<Self, OrderedStoreError> {
+        let mut cursor = Self {
+            env,
+            index_to_key_db,
+            main_db,
+            cache: VecDeque::new(),
+            range: range.unwrap_or_else(|| (..).into()), // default to unbounded range
+        };
+
+        // Load initial values into the cache
+        if let Err(err) = cursor.reload_cache() {
+            error!("Failed to load iterator's initial cache: {}", err);
+        }
+
+        Ok(cursor)
+    }
+
+    fn reload_cache(&mut self) -> Result<(), String> {
+        let txn = lmdb::ReadTransaction::new(self.env.clone()).map_err(|err| err.to_string())?;
```

Comment:
Is there no way to keep the transaction for the iterator?

Reply:
Fixed